### PR TITLE
feat(mirror): add a transactions per second option

### PR DESCRIPTION
**Pull request by marcelo-gonzalez**
This adds an option to control the traffic load by specifying how many transactions per second to send. This still sends the same set of transactions, but iterates through them and selects an amount to send that will get us close to the tps number we're looking for, rather than just sending a whole block's worth at a time

This PR is complicated by the way things work in this mirror tool currently, and it would be great to find a way to simplify things and delete code in the future. But for now sadly things have to get even a little more complicated to implement this, since there's some stuff that could go wrong with the way nonces are handled

### CODE REVIEW

File: tools/mirror/src/lib.rs [hunk 1]
```diff
@@ -212,18 +208,74 @@ fn delete_pending_outcome(db: &DB, id: &CryptoHash) -> anyhow::Result<()> {
     )?)
 }
 
-fn set_last_source_height(db: &DB, height: BlockHeight) -> anyhow::Result<()> {
-    // TODO: we should instead save something like the
-    // (block_height, shard_id, idx_in_chunk) of the last
-    // transaction sent. Currently we set last_source_height after
-    // sending all of the transactions in that chunk, so if we get
+#[derive(Debug, BorshDeserialize, BorshSerialize)]
+enum TxBookmark {
```
Comment (tayfunelmas) [inline tools/mirror/src/lib.rs:212]: can you add comments to the structs and fields about what they represent and/or how they are used?

File: tools/mirror/src/lib.rs [hunk 2]
```diff
@@ -234,6 +286,22 @@ fn get_last_source_height(db: &DB) -> anyhow::Result<Option<BlockHeight>> {
         .map(|v| BlockHeight::try_from_slice(&v).unwrap()))
 }
 
+fn get_first_accounts_created(db: &DB) -> anyhow::Result<bool> {
```
Comment (tayfunelmas) [inline tools/mirror/src/lib.rs:289]: what does creating first accounts mean? is it about the initial setup? can you add comments to clarify these?

File: tools/mirror/src/chain_tracker.rs [hunk 3]
```diff
@@ -121,7 +121,53 @@ struct NonceInfo {
...
+    TPS(TxBatchConfig),
+}
+
+impl TrafficCadence {
+    fn new(tx_batch_interval: Option<Duration>, tps: Option<u32>) -> Self {
+        assert!(tx_batch_interval.is_none() || tps.is_none());
+        if let Some(tps) = tps {
+            let c = if tps <= MAX_BATCHES_PER_SEC {
+                TxBatchConfig {
+                    batch_count: 1,
```
Comment (tayfunelmas) [inline tools/mirror/src/chain_tracker.rs:151]: I think comment for batch_count would help, I do not understand why the batch is 1 here, do you mean everything within a second will be sent together or something else?

File: tools/mirror/src/lib.rs [hunk 4]
```diff
@@ -212,18 +208,74 @@ fn delete_pending_outcome(db: &DB, id: &CryptoHash) -> anyhow::Result<()> {
...
+            db.delete_cf(db.cf_handle(DBCol::Misc.name()).unwrap(), "last_source_height")?;
+            Ok(())
+        }
+        None => Ok(()),
+    }
+}
+
+fn get_traffic_checkpoint(db: &DB) -> anyhow::Result<Option<TrafficCheckpoint>> {
+    Ok(db
+        .get_cf(db.cf_handle(DBCol::Misc.name()).unwrap(), "traffic_checkpoint")?
```
Comment (tayfunelmas) [inline tools/mirror/src/lib.rs:266]: can you define these constants (eg. traffic_checkpoint and last_source_height at the top of the file as the DB keys and explain what they are supposed to contain?

**Review by tayfunelmas — CHANGES_REQUESTED** [review]
I put some comments about adding comments to the code because it is hard to parse what is going on. In addition, can you explain the design shortly in the PR description?

**Comment by marcelo-gonzalez** [comment]
going to finish testing this, and I think there might be some extra bugs in here that need to be fixed, then i'll add all these comments, since I agree it is very light on explanations here...
